go/consensus/tendermint/apps/beacon/backend_scrape.go

```diff
@@ -0,0 +1,819 @@
+		// Should never happen.
+		return fmt.Errorf("beacon: unexpected reveal tx")
+	}
+
+	// Process the commit (CPU INTENSIVE).
+	if err := scrapeState.Instance.OnReveal(revealTx.Reveal); err != nil {
+		return fmt.Errorf("beacon: failed to process reveal tx: %w", err)
+	}
+
+	return nil
+}
+
+func (impl *backendSCRAPE) doSetEpochTx(
+	ctx *api.Context,
+	state *beaconState.MutableState,
+	tx *transaction.Transaction,
+) error {
+	// Handle the mock backend SetEpoch transaction.
+	now, _, err := state.GetEpoch(ctx)
+	if err != nil {
+		return fmt.Errorf("beacon: failed to get current epoch: %w", err)
+	}
+
+	var epoch beacon.EpochTime
+	if err = cbor.Unmarshal(tx.Body, &epoch); err != nil {
+		return fmt.Errorf("beacon: failed to deserialize set epoch tx: %w", err)
+	}
+
+	// Ensure there is no SetEpoch call in progress.
+	pendingMockEpoch, err := state.SCRAPEPendingMockEpoch(ctx)
+	if err != nil {
+		return fmt.Errorf("beacon: failed to query mock epoch state: %w", err)
+	}
+	if pendingMockEpoch != nil {
+		// Unless the requested explicit epoch happens to be pending.
+		if *pendingMockEpoch == epoch {
+			return nil
+		}
+		return fmt.Errorf("beacon: explicit epoch transition already pending")
+	}
+
+	if epoch <= now {
+		// Constructing closed timelike curves is left for civilizations
+		// that have mastered spacetime metric engineering such as
+		// the Xeelee, and has no place in a trivial blockchain project.
+		ctx.Logger().Error("explicit epoch transition does not advance time",
+			"epoch", now,
+			"new_epoch", epoch,
+		)
+		return fmt.Errorf("beacon: explicit epoch chronology violation")
+	}
+
+	if err = state.SetSCRAPEPendingMockEpoch(ctx, epoch); err != nil {
+		return fmt.Errorf("beacon: failed to set pending mock epoch: %w", err)
+	}
+
+	ctx.Logger().Info("scheduling explicit epoch transition on round completion",
+		"epoch", epoch,
+	)
+
+	return nil
+}
+
+func (impl *backendSCRAPE) doEmitSCRAPEEvent(ctx *api.Context, scrapeState *beacon.SCRAPEState) {
+	var event beacon.SCRAPEEvent
+	event.FromState(scrapeState)
+
+	ctx.EmitEvent(api.NewEventBuilder(impl.app.Name()).Attribute(
+		KeySCRAPERound,
+		cbor.Marshal(event),
+	))
+}
+
+func (impl *backendSCRAPE) appendFailures(scrapeState *beacon.SCRAPEState, failures []signature.PublicKey) {
+	if len(failures) == 0 {
+		return
+	}
+	if scrapeState.BadParticipants == nil {
+		scrapeState.BadParticipants = make(map[signature.PublicKey]bool)
+	}
+	for _, id := range failures {
+		scrapeState.BadParticipants[id] = true
+	}
+}
+
+func (impl *backendSCRAPE) updateNodesSnapshot(ctx *api.Context, state *beaconState.MutableState) error {
+	registryState := registryState.NewMutableState(ctx.State())
+	nodes, err := registryState.Nodes(ctx)
+	if err != nil {
+		return fmt.Errorf("beacon: failed to query node list: %w", err)
+	}
+
+	var ids []signature.PublicKey
+	for _, node := range nodes {
+		ids = append(ids, node.ID)
+	}
+
+	if err = state.SetSCRAPENodesSnapshot(ctx, ids); err != nil {
```

Comment:
It's super annoying to do if we want to continue to support the mock epoch, because it is not guaranteed to know `that specific epoch` when this happens.

Reply:
So while I took a stab at addressing this, I'm still against changing it as it complicates the code and breaks the e2e tests (timeouts).  I'm not sure if it's a bug in the changes or just "it takes even longer now", and I'm beyond reluctant to keep increasing the timeout (this stuff takes forever already).

Reply:
Ok, feel free to revert, we can look at this later.

Reply:
Seems to work now, after I debugged it a bit.